src/event/workqueue.c

```diff
@@ -0,0 +1,287 @@
+#endif
+	dispatch_workq_manager_t mgr = _dispatch_workq_get_default_manager();
+	int tid = syscall(SYS_gettid);
+	_dispatch_unfair_lock_lock(&mgr->registered_worker_lock);
+	for (int i=0; i<mgr->num_registered_workers; i++) {
+		if (mgr->registered_workers[i] == tid) {
+			int last = mgr->num_registered_workers - 1;
+			mgr->registered_workers[i] = mgr->registered_workers[last];
+			mgr->registered_workers[last] = 0;
+			mgr->num_registered_workers--;
+			break;
+		}
+	}
+	_dispatch_unfair_lock_unlock(&mgr->registered_worker_lock);
+}
+
+
+/*
+ * For each pid that is a registered worker, read /proc/[pid]/stat
+ * to get a count of the number of them that are actually runnable.
+ * See the proc(5) man page for the format of the contents of /proc/[pid]/stat
+ */
+static void
+_dispatch_workq_count_runnable_workers(dispatch_workq_manager_t mgr)
+{
+	char path[128];
+	char buf[4096];
+	int running_count = 0;
+
+	memset(buf, 0, sizeof(buf));
+
+	_dispatch_unfair_lock_lock(&mgr->registered_worker_lock);
+
+	for (int i=0; i<mgr->num_registered_workers; i++) {
+		pid_t worker_pid = mgr->registered_workers[i];
+		int fd;
+		size_t bytes_read = -1;
+
+		int r = snprintf(path, sizeof(path), "/proc/%d/stat", worker_pid);
+		(void)dispatch_assume(r > 0 && r < sizeof(path));
+
+		fd = open(path, O_RDONLY | O_NONBLOCK);
+		if (unlikely(fd == -1)) {
+			// Unable to open file.
+			// Must mean worker exited uncleanly (without executing _dispatch_worker_unregister())
+			// Clean up by removing pid and decrementing number of registered workers
+			_dispatch_debug("workq: Unable to open /proc/%d/stat; removing worker from monitoring list", worker_pid);
+			int last = mgr->num_registered_workers-1;
+			mgr->registered_workers[i] = mgr->registered_workers[last];
+			mgr->registered_workers[last] = 0;
+			mgr->num_registered_workers--;
+		} else {
+			bytes_read = read(fd, buf, sizeof(buf));
+			(void)close(fd);
+		}
+
+		if (bytes_read > 0) {
+			char state;
+			if (sscanf(buf, "%*d %*s %c", &state) == 1) {
+				// _dispatch_debug("workq: The state of worker %d is %c\n", worker_pid, state);
+				if (state == 'R') {
+					running_count++;
+				}
+			} else {
+				_dispatch_debug("workq: Failed to scan state for worker %d", worker_pid);
+			}
+			memset(buf, 0, bytes_read);
+		} else {
+			_dispatch_debug("workq: Failed to read %s", path);
+		}
+	}
+
+	mgr->runnable_workers = running_count;
+
+	_dispatch_unfair_lock_unlock(&mgr->registered_worker_lock);
+}
+
+static void
+_dispatch_workq_monitor_thread_pool(void *context DISPATCH_UNUSED)
+{
+	dispatch_workq_manager_t mgr = _dispatch_workq_get_default_manager();
+	dispatch_queue_t dq = _dispatch_workq_get_default_root_queue();
+	bool work_available = _dispatch_workq_root_queue_has_work(dq);
+
+	if (!work_available) {
+		_dispatch_debug("workq: %s is empty; doing nothing", dq->dq_label);
+		return;
+	}
+
+	_dispatch_workq_count_runnable_workers(mgr);
+	int32_t count = _dispatch_pthread_root_queue_thread_pool_size(dq);
```

Comment:
(it goes with my general review remark about the fact you need to count threads in a pre-posted way anyway to avoid in-flight thread explosion)